src/python/pants/backend/docker/util_rules/docker_build_context.py

```diff
@@ -127,12 +128,12 @@ async def create_docker_build_context(
     # Get all dependencies for the root target.
     root_dependencies = await Get(Targets, DependenciesRequest(docker_image.get(Dependencies)))
 
-    # Get all sources from the root dependencies (i.e. files).
-    file_sources_request = Get(
+    # Get all sources from the root dependencies (i.e. files and shell sources).
+    sources_request = Get(
         SourceFiles,
         SourceFilesRequest(
             sources_fields=[tgt.get(SourcesField) for tgt in root_dependencies],
-            for_sources_types=(FileSourceField,),
+            for_sources_types=(FileSourceField, ShellSourceField),
```

Comment:
Hm... tricky. Because you want codegen to work here (in general), you can't specify "all source files" by not passing the `for_sources_types` argument.

I guess that that makes sense though... it probably wouldn't make sense to depend on gRPC generated `PythonSources` here (for example): rather, some binary target that they were included in, and which would be converted into loose files.

So I guess that that suggests that another way to do this would be to add rules that would "codegen" `files` from `shell_sources`. An advantage to that is that you could "relocate" `shell_sources` with `relocated_files` before including them here?

cc @Eric-Arellano , since this is the kind of question to think about before stabilizing the plugin API.

Reply:
I actually think that codegen'ing files from `shell_sources` might make more sense: interested in your thoughts.

Reply:
I like that idea. As that makes this rule agnostic to where the files come from, and other plugins may codegen from any source to get it into a build context. 

I'll go ahead and make that change tomorrow.

Reply:
I'm not sure what would happen though with the Shell backend when it says to include both Files and ShellSources? I guess it will produce two copies of the exact same files, and so the `Digest` will deduplicate them both.

--

A weird side-effect of this is that the `export-codegen` rule will now write all shell sources into the `dist/codegen` folder...it was already weird that `relocated_files` would do that too.

To address that, maybe we should update `export-codegen` to not export certain codegen rules like this that aren't proper codegen? We can use a class property to indicate which requests should be exported or not. Wdyt?

Reply:
> I'm not sure what would happen though with the Shell backend when it says to include both Files and ShellSources? I guess it will produce two copies of the exact same files, and so the Digest will deduplicate them both.

Good point. But it would be a bit odd for a `shell_library` to request `files`, since a `shell_library` is not itself runnable? It seems like that depends on what `SourcesRequest` consuming a `shell_library` makes. An `experimental_shell_command` could though, so that's probably what you meant.

> To address that, maybe we should update export-codegen to not export certain codegen rules like this that aren't proper codegen? We can use a class property to indicate which requests should be exported or not. Wdyt?

Possibly. "codegen" has been a misnomer for a while now though... it's more like a "field conversion `@rule`" or something. So yea, how we know that something doesn't correspond to a file on disk could change to some other property maybe.

Reply:
Ah, true. So in that case, perhaps a dedicated union type for hydrating sources to the build context is better, and leave codegen out?

Reply:
> Good point. But it would be a bit odd for a shell_library to request files, since a shell_library is not itself runnable?

Shfmt, Shellcheck, and Shunit2 all need to request `ShellSingleSourceField`.

> So in that case, perhaps a dedicated union type for hydrating sources to the build context is better, and leave codegen out?

I think we still use "codegen" as implemented. Just have better support for "This isn't reallyyyyy codegen like Protobuf. It's (ab)using the mechanism."

Reply:
> "...It's (ab)using the mechanism."

I'm not sure that I agree on this part... it seems like there is a clear desire-path for being able to convert source fields for all kinds of reasons, to the point where making the existing mechanism generic makes sense. It's a turning out to be a fundamental interop tool.

Reply:
Given that a user pointed out they need to include `PythonSources` here, I think using codegen to convert `ShellSources` into `FilesSources` is the wrong approach. I doubt we want to add that codegen rule for `PythonSources` because of source roots.

Instead, I think what we probably want is to loosen the restrictions on codegen so that you are able to say `for_sources_types=(SourcesField,), enable_codegen=True` and it will trigger _all_ relevant codegen. This means that a single Protobuf file might generate Java and Python etc. Callers can still use `for_sources_types` to only lazily generate what is relevant, e.g. Python only asks for Python.

The only reason I think I added that restriction when adding codegen was because we wanted codegen to be as lazy as possible. I think the restriction was more than anything meant to assist rule authors to not do the suboptimal thing. Turns out that is causing too many issues.

Wdyt?

Reply:
> what are the relevant codegen rules in that case?

The way the code currently works is that when it encounters a `SourcesField` subclass, we first check if it's a direct subclass of one of the `for_sources_types`. So if you have `for_sources_types=(PythonSourceField,)` and then a `PythonTestSourceField` instance, that's easy. Then, if that fails, we inspect the `UnionMembership` to see if there is a codegen rule that goes from the instance -> one of the expected types.

The issue is that we error if there are >1 rules. I'm proposing that we remove that restriction. If the sources instance can be converted into >1 of the expected types, cool. Generate it all.

Although, I'm not certain still that's a good change. If we set `for_sources_types=(SourcesField,), enable_codegen=True` and you have Protobuf sources, that would result in including the original `.proto` files _and_ generating Python _and_ generating any other languages like Go and Java. The `.proto` files are a subclass of `SourcesField`, and the languages can be generated into subclasses of `SourcesField`.

Reply:
Ah, ok, those releavant codegens. Thanks. Yeah, I've been on a dinner break, will see if I can make some more headway with my approach, see if it's feasible, as it wouldn't require such a change to the core of codegen.

Reply:
I've pushed a fix, where I have a new dedicated output type for the codegen, and an input type that "accepts" all sources field types that we want to consume into the docker build context, simply by registering them. 

I think it turned out really well, with only 1 minor tweak to the export codegen goal, so we can opt-out of that.

cc @stuhood @Eric-Arellano ;)